pkg/daemon/ceph/osd/kms/envs.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kms
+
+import (
+	"os"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	v1 "k8s.io/api/core/v1"
+)
+
+var (
+	knownKMSPrefix = []string{"VAULT_"}
+)
+
+// VaultTokenEnvVarFromSecret returns the kms token secret value as an env var
+func vaultTokenEnvVarFromSecret(tokenSecretName string) v1.EnvVar {
+	return v1.EnvVar{
+		Name: api.EnvVaultToken,
+		ValueFrom: &v1.EnvVarSource{
+			SecretKeyRef: &v1.SecretKeySelector{
+				LocalObjectReference: v1.LocalObjectReference{
+					Name: tokenSecretName,
+				},
+				Key: KMSTokenSecretNameKey,
+			},
+		},
+	}
+}
+
+// vaultTLSEnvVarFromSecret translates TLS env var which are set to k8s secret name to their actual path on the fs once mounted as volume
+// See: TLSSecretVolumeAndMount() for more details
+func vaultTLSEnvVarFromSecret(kmsConfig map[string]string) []v1.EnvVar {
+	vaultTLSEnvVar := []v1.EnvVar{}
+
+	for _, tlsOption := range vaultTLSConnectionDetails {
+		tlsSecretName := GetParam(kmsConfig, tlsOption)
+		if tlsSecretName != "" {
+			vaultTLSEnvVar = append(vaultTLSEnvVar, v1.EnvVar{Name: tlsOption, Value: path.Join(EtcVaultDir, tlsSecretPath(tlsOption))})
+		}
+	}
+
+	return vaultTLSEnvVar
+}
+
+// VaultConfigToEnvVar populates the kms config as env variables
+func VaultConfigToEnvVar(kmsConfig map[string]string, tokenSecretName string) []v1.EnvVar {
+	envs := []v1.EnvVar{}
+	for k, v := range kmsConfig {
+		// Skip TLS and token env var to avoid env being set multiple times
+		toSkip := append(vaultTLSConnectionDetails, api.EnvVaultToken)
+		if client.StringInSlice(k, toSkip) {
+			continue
+		}
+		envs = append(envs, v1.EnvVar{Name: k, Value: v})
+	}
+
+	// Add the VAULT_TOKEN
+	envs = append(envs, vaultTokenEnvVarFromSecret(tokenSecretName))
+
+	// Add TLS env if any
+	envs = append(envs, vaultTLSEnvVarFromSecret(kmsConfig)...)
+
+	logger.Debugf("kms envs are %v", envs)
+
+	return envs
+}
+
+// ConfigEnvsToMapString returns all the env variables in map from a known KMS
+func ConfigEnvsToMapString() map[string]string {
```

Comment:
If the envs were in the cluster CR instead of a separate configmap, you could already have them as a `map[string]string` from types.go

Reply:
I know... but the CM is much more flexible, let's discuss this further.